### PR TITLE
feat(renderer): PBR closure v2 — versioned Evaluate/Sample/Pdf contract, energy compensation, near-mirror fix

### LINKED ISSUE #975: Renderer: PBR closure v2 — versioned Evaluate/Sample/Pdf contract, energy compensation, near-mirror fix
## Problem

OloEngine's CPU reference path tracer deliberately mirrors `OloEditor/assets/shaders/include/PBRCommon.glsl`, which is the right policy for renderer-parity testing. Issue #706 already landed a large part of a modern specular stack there — GGX VNDF importance sampling (`sampleGGXVNDF`), the Heitz G2/G1 sample weight (`ggxVNDFWeight`), height-correlated Smith visibility, a shared blue-noise sampler, plus white-furnace tests (`PathTracerFurnaceTest.cpp`) and CPU/GPU parity tests (`ReferenceBRDFGpuParityTest.cpp`).

What is **not** yet in place is the part a GPU path tracer and a ReSTIR stack actually depend on:

* **No versioned material model.** Changing the closure today silently changes every existing scene and golden. There is no way to introduce a corrected closure without a migration cliff.
* **No `Evaluate` / `Sample` / `Pdf` triple as one contract.** BRDF evaluation exists as renderer code and VNDF sampling exists as shader code, but nothing pins them to a single agreeing density. Path tracing needs all three to be mathematically consistent; MIS needs the density they agree on.
* **Near-mirror lobe collapse.** `distributionGGX` divides by `max(denom, EPSILON)` with `EPSILON = 1e-4` (`PBRCommon.glsl:151`). As roughness approaches 0 the numerator goes to zero while the clamped denominator does not, so the lobe collapses toward black instead of approaching a mirror.
* **Two geometry terms coexist.** The direct-light path still uses the Schlick-GGX `k` approximation (`PBRCommon.glsl:222`) while the VNDF path uses height-correlated Smith. Both are defensible; having both under one "the BRDF" label is not.
* **No energy compensation.** There is no multiple-scattering / energy-loss correction anywhere in the tree, so rough metals lose energy that a path tracer will make obvious.
* The current Fresnel/diffuse split is common in real-time renderers but mildly non-reciprocal.

Full path tracing amplifies inconsistencies between `Evaluate`, `Sample`, and `Pdf`. ReSTIR then reuses those samples across pixels and frames, making a superficially plausible but biased material implementation much harder to diagnose.

## Goal

Introduce a **versioned PBR closure v2** that is shared conceptually — and parity-tested mechanically — across raster lighting, hybrid ray tracing, the CPU reference tracer (#709), a future GPU path tracer, and future ReSTIR passes.

The core contract should expose equivalents of:

```cpp
BSDFEval   Evaluate(const Surface& s, vec3 wo, vec3 wi);
BSDFSample Sample  (const Surface& s, vec3 wo, Sample3D xi);
float      Pdf     (const Surface& s, vec3 wo, vec3 wi);
```

For every non-delta sample, `Sample().Pdf`, `Pdf()`, and the density used by MIS must agree.

### 1. Version rather than silently mutate

Add an explicit material/renderer model version, for example:

```cpp
enum class PBRModel : u8
{
    Legacy,
    ClosureV2
};
```

Existing scenes/goldens remain on `Legacy` until migrated deliberately. New benchmark scenes may opt into v2 immediately; switching the project default happens only after parity and visual review. This is the single most important deliverable in the issue — everything else is safe to land once the switch exists.

### 2. Define the v2 base closure

Reuse what #706 already built; the new work is the surrounding contract, not a rewrite of the sampler.

* Lambertian or Burley-style diffuse with an explicitly documented energy split.
* GGX/Trowbridge-Reitz microfacet reflection.
* **Reuse** the existing Heitz VNDF sampler and G2/G1 weight rather than writing a second one.
* **Reuse** the existing height-correlated Smith masking-shadowing, and retire the parallel Schlick-GGX `k` path for v2 surfaces (Legacy keeps it).
* Dielectric/metallic Fresnel with a single documented `F0` convention.
* A documented perceptual-roughness mapping, e.g. `alpha = max(r^2, alpha_min)`.
* **New:** stable near-mirror handling that does not collapse because of a denominator clamp — clamp `alpha`, not the denominator.
* **New:** multiple-scattering energy compensation, or another documented energy-loss correction.
* Consistent geometric-normal versus shading-normal conventions.

Keep the first version intentionally smaller than a complete Disney/MaterialX closure system.

### 3. One authoritative specification

Avoid indefinitely maintaining unrelated hand-written formulas in C++ and GLSL. Acceptable strategies:

* generate a C++ and shader include from a small authoritative description; or
* retain explicit C++/GLSL implementations but pin every function with cross-language parity tests and a versioned formula ledger.

`ReferenceBRDFGpuParityTest.cpp` is the existing mechanism for the second option and should be extended rather than replaced. Record the choice in a short design note — it becomes the contract for all later transport estimators.

### 4. Validation — extend, do not duplicate

`PathTracerFurnaceTest.cpp` (5 tests) and `ReferenceBRDFGpuParityTest.cpp` (3 tests) already exist. Extend them to cover v2:

* white-furnace energy behavior over roughness/metallicity/view-angle grids **including the new energy compensation**;
* hemispherical reflectance not exceeding the documented tolerance;
* reciprocity where the selected closure is expected to be reciprocal;
* finite, non-negative BRDF and PDF values;
* **low-roughness convergence toward a narrow specular lobe** (the regression test for the clamp fix);
* `Evaluate`/`Sample`/`Pdf` consistency by Monte Carlo integration (new — nothing covers this today);
* deterministic edge cases at grazing angles and zero/one roughness;
* GPU parity for the same sample tuples through `PBRCommon.glsl`.

## Acceptance criteria

* `Legacy` and `ClosureV2` are explicit, versioned choices; existing assets do not change appearance silently.
* Raster, CPU reference PT, and GPU shader tests agree for the v2 evaluation contract within documented floating-point tolerances.
* The v2 sampler and PDF pass Monte Carlo consistency tests over representative materials.
* The near-zero-roughness case no longer collapses to a dark/vanishing lobe, with a test that fails against today's code.
* Energy compensation is present and covered by a furnace test that fails without it.
* v2 uses exactly one geometry term; the Schlick-GGX path remains reachable only through `Legacy`.
* The material laboratory benchmark captures v2 roughness/metallicity sweeps from multiple view angles.
* The implementation documents all deliberate clamps, approximations, and biased production modes.
* Shader hot reload and both raster RHIs continue to work.

## Explicitly out of scope

* Re-implementing VNDF sampling, the G2/G1 weight, or height-correlated Smith — #706 already landed those.
* Spectral rendering.
* Subsurface scattering, skin, hair/fur, layered car paint, thin-film interference, or measured BRDFs.
* Full dielectric transmission and participating media in the first v2 slice.
* Replacing every specialized material — water, foliage, particles — with the base closure.

## References

* Heitz, *Sampling the GGX Distribution of Visible Normals*.
* Kulla and Conty, *Revisiting Physically Based Shading at Imageworks* (energy compensation).
* #706 — the existing VNDF / blue-noise / temporal-resolve work.
* #709 — CPU reference path tracer; `ReferenceBRDF.h` parity comments and `PBRCommon.glsl`.

## Score
```olo-score
capability: 8
craft: 8
stability: 8
decay: 5
effort: 8
confidence: 0.8
learning: 8
fun: 8
kano: table-stakes
blocked_by: []
blocks: []
```
<sub>Rated per [issue-scoring](https://github.com/drsnuggles8/OloEngineBase/blob/master/docs/process/issue-scoring.md) · score = confidence × (capability + craft + stability + decay) / effort, derived by the picker.</sub>

### PR DESCRIPTION
**Pull request by drsnuggles8**
## What this is

Issue #975: the material closure becomes an explicit, versioned choice, so a corrected
BRDF can ship without silently changing a single existing pixel.

**`PBRModel { Legacy = 0, ClosureV2 = 1 }`** is a per-material field (`Material::SetPBRModel`,
the `MaterialType` precedent), default **Legacy — frozen bit-for-bit**. It travels
Material → `PODMaterialData` (+ its `==` cache key) → `PBRMaterialUBO` (the former `Pad2`
lane; still 144 B) → `u_PBRModel` in all 8 GLSL block sites → forward lighting via a
`calculateLightContribution`/`fplusEvaluateTileLights` overload, and on Deferred via **bit 1
of the G-Buffer RT2 alpha flags lane** (bit 0 stays the unlit flag; Legacy still writes 0.0,
so existing G-Buffer bytes are unchanged). Scene YAML writes the field only when non-Legacy
and **Rejects** out-of-range values; save-games append it version-gated (24→25); the
inspector gets a combo; Lua gets `pbrModel`.

**ClosureV2** differs from Legacy in exactly four documented ways (PBRCommon.glsl "PBR
CLOSURE V2" section + ADR 0016):
1. **One geometry term** — height-correlated Smith visibility. The Schlick-GGX `k` remap
   stays reachable only through Legacy (acceptance criterion).
2. **Near-mirror: clamp alpha, not the denominator** — `alpha = clamp(r, 0.04, 1)²`, NDF
   denominator unclamped. The lobe brightens toward a mirror instead of collapsing.
3. **Kulla-Conty multiple-scattering energy compensation**, reciprocal form, driven by a
   generated 16×16 Ess/E_avg loss table. The entries are the expected VNDF weight
   E[G2/G1] — computed with the v2 sampler's own estimator identity — and at F_avg = 1
   the furnace closes to 1 analytically.
4. **One D serves Evaluate, Sample and Pdf** — the consistency a path tracer/ReSTIR needs,
   structurally impossible for Legacy (whose evaluation-side clamp must never enter a
   density).

**The contract triple** lives in `PBRClosureBSDF.h` (`BSDF::Evaluate/Sample/Pdf`,
dispatching on `ReferenceMaterial::Model`; the Legacy bodies moved verbatim from
`PathTracer.cpp` — the bit-identical render hashes pin the move) and in GLSL as
`closureV2Evaluate/SampleBRDF/Pdf`. Specification strategy per the issue's option 2:
explicit twin implementations pinned by cross-language parity tests + the formula ledger
(ADR 0016 §3). The one generated artifact is the energy-table pair, emitted as **identical
half-packed hex words** into `GgxEnergyTables.h` and `PBRClosureV2Energy.glsl` — packed
because a plain `const float[256]` passed glslc but **failed NVIDIA's GL linker**
(`C5025: lvalue in assignment too complex`) once inlined at PBR_MultiLight's three lighting
call sites; the postmortem is now `glsl-shaders.md` §8a.

## Acceptance criteria → evidence

| Criterion | Evidence |
|---|---|
| Legacy/ClosureV2 explicit, no silent change | `ClosureV2.LegacyDispatchIsUntouched` (bit-identity vs `CookTorranceBRDF`), `PathTracerCornellBox.RendersAreBitIdentical` (hashes unchanged), zero golden churn, Legacy G-Buffer bytes unchanged |
| Raster / CPU reference / GPU agree on v2 | `ClosureV2GpuParity.V2EvaluateMatchesCppTwin` + `V2PdfMatchesCppTwin` (1%/1e-4 over a 128×128 grid, run on real GPU) — the density is now cross-language-pinned, which Legacy never had |
| MC consistency of sampler + PDF | `ClosureV2Consistency.*` — Sample/Pdf bit-equal; importance estimate vs disjoint-code quadrature; density normalization incl. the below-horizon VNDF mass; end-to-end traced v2 furnace |
| Near-mirror no longer collapses, test fails on today's code | `ClosureV2.NearMirrorLobeNoLongerCollapses` — v2 albedo ≥ 0.97 where the Legacy closure measures < 0.2 (paired negative) |
| Energy compensation + furnace test that fails without it | `ClosureV2.FurnaceClosesWithCompensationAndFailsWithoutIt` — [0.95, 1.05] with the term, < 0.90 without at r ≥ 0.6 |
| v2 uses exactly one geometry term | `closureV2Evaluate` composes only `visibilitySmithGGXCorrelated`; pinned by the parity probes |
| Material laboratory, multi-angle sweeps | `MaterialLabVisualEvidence.RoughnessMetallicSweepAcrossViewAngles` — 6 committed PNGs (2 models × 3 view angles), differential contracts, frames eyeballed |
| Documented clamps/approximations/biased modes | ADR 0016 §5 (the honest list) |
| Hot reload + both raster paths | live editor session below; `ClosureV2DeferredParityScene.ModelSelectorIsLiveOnBothRasterPaths` pins Forward AND Deferred |

## Verification

- **Full suite: 6897 tests, 0 failures** (Debug, RTX 4090, GL 4.6; skips are the documented
  GPU-less/steam-stub set).
- **Live editor over MCP, both paths**: authored `PBRClosureV2Test.olo` (five Legacy/V2
  sphere pairs; also exercises the scene-YAML `PBRModel` round-trip end-to-end), opened it
  in the running editor, captured Forward and Deferred frames, and probed G-Buffer texels
  (`olo_render_probe_pixel`) to verify the flags lane byte-for-byte. TEST_SCENES.md
  documents the pass/fail reading.
- **The live loop caught a real bug the headless suite could not**: the deferred path
  decoded the model bit and then never passed it into the lighting calls — every v2
  material shaded Legacy on Deferred while Forward was perfect. Fixed, and pinned by the
  new `ClosureV2DeferredParityEvidenceTest`, whose failure mode was verified by
  re-introducing the bug (identical peaks, exact diagnostic) before restoring the fix.
- The NVIDIA `C5025` link failure was likewise invisible to glslc and to single-shader GPU
  probes; it surfaced as 45 assorted visual-test failures, was bisected to the const-array
  emission, and is fixed by the packed tables (verified: all 45 back to green).

## Self-review (multi-angle, high effort) — all 10 findings dispositioned

Fixed before this PR: the **TileRendererComponent save-game serializer** now carries
`PBRModel` per tile material (the second full-Material save site; caught as a CONFIRMED
silent-drop); the **energy-lookup contract** (call sites now pass authored roughness as
documented, both languages); the four hand-duplicated `<= 1` range literals now share
`kPBRModelCount` (with the deferred-lane ceiling documented beside it); the four
pasted G-Buffer flag writes route through one `oloEncodeGBufferPbrFlags` helper (the
layout's single executable home); the `closureV2SampleBRDF` failure-convention comment
no longer claims 0/0 is safe; and assorted stale comments.

Dismissed with documentation (each carries an in-code comment pointing at ADR 0016 §5):
the **resolved-MSAA flags-lane averaging** limitation (PLAUSIBLE; reachable only via two
non-default session-local toggles, and the pre-#975 unlit flag had the milder variant of
the same defect — the real fix is excluding the lane from the averaged resolve, a
follow-up); the **imported-material codec** and **asset-preview** lanes staying Legacy
(importers can't author v2; the preview shader is a standalone pre-dispatch closure
copy); the **C# `OLO_PROPERTY`** exemption (matches the un-annotated `AlphaMode`
precedent, recorded at the component). Deferred as noted follow-ups: hoisting the
per-pixel-invariant Kulla-Conty terms out of the light loops, a constexpr-decoded CPU
table, checking in the table generator as a tool, and sharing the test-side
Hammersley/probe-harness helpers.

## Deliberately out of scope (recorded in ADR 0016 §5)

- Sphere-area lights shade v2 materials through the Legacy representative-point evaluator.
- IBL/ambient stays split-sum with no multi-scatter term in this slice.
- No global/scene default knob — the project default is the constructor default; flipping
  any default is a separate decision with a golden rebake attached.
- `.omat` MaterialAsset YAML and the imported-material binary codec don't carry the field
  (they persist no enum at all today; imported materials are Legacy in this slice).
- No C# `OLO_PROPERTY` binding (Lua has `pbrModel`); no MCP field-registry entry (the
  field lives inside `Material`, not on the component).

## Review guide

**Where I'd look hardest**

1. `OloEditor/assets/shaders/include/DeferredLightingShared.glsl` — the flags-lane decode
   (`bit 0 unlit, bit 1 model`) and the two lighting calls now taking `pbrModel`. This is
   the seam that was silently wrong once already; the new parity evidence test is its pin.
2. `OloEngine/src/OloEngine/Renderer/PathTracing/PBRClosureBSDF.h` — the claim that the
   Legacy branch is a *verbatim* move of the former PathTracer-internal bodies (the
   Cornell bit-identity tests enforce it, but the equivalence argument is worth a read).
3. `PBRCommon.glsl` "PBR CLOSURE V2" — `closureV2Pdf` must be exactly the density
   `closureV2SampleBRDF` draws from; the below-horizon convention comment explains the
   deliberate asymmetry the consistency test accounts for.

**What I verified, and how** — the table above; plus the paired-negative discipline on the
three new failure-mode tests (furnace-without-compensation, Legacy near-mirror collapse,
deferred-lane-dead), each shown to fail on the code it guards against.

**Least confident about** — the faint mach-banding on roughness-1.0 cells in the
MaterialLab captures from the 16-bin table's derivative discontinuities under bilinear
interpolation. Cosmetic, v2-only, bounded by the documented table resolution; a 32-bin
table or Catmull-Rom lookup is the follow-up if it ever shows on real content.

**Deliberately not tested** — MaterialComponent YAML round-trip headless (documented SEH
constraint in ComponentRoundTripTest.cpp; exercised live via PBRClosureV2Test.olo
instead). Vulkan-backend v2 rendering (the UBO lane is backend-neutral and the Vulkan
tier compiles the same GLSL, but no live Vulkan frame was captured — the driver has no
RHI passthrough for `attach`).

Closes #975

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added optional Closure V2 PBR shading for improved near-mirror highlights and rough-metal energy response.
  * Added editor and scripting controls to switch materials between Legacy and Closure V2.
  * Preserved material selections across Forward and Deferred rendering, scenes, and saves.
  * Added a PBR closure comparison test scene.

* **Documentation**
  * Documented versioned PBR closure behavior and validation criteria.

* **Tests**
  * Added CPU, GPU, rendering-parity, serialization, and visual-evidence coverage.

### CODE REVIEW

**Comment by drsnuggles8** [comment]
## 🤖 Self-review @ `7e6509126d7a660bcf364021f07f7a8657f972fc`
Reviewed the full working-tree diff at **high** effort (multi-angle: line-by-line, removed-behavior, cross-file trace, reuse, simplification, efficiency, altitude, conventions; every candidate verified against code).
- **Findings:** 10 · **Fixed:** TileRendererComponent save-game now carries PBRModel (the second full-Material site); energy-lookup call sites honor the authored-roughness contract (both languages); `kPBRModelCount` replaces four hand-duplicated range literals; the four pasted G-Buffer flag writes route through one `oloEncodeGBufferPbrFlags` home; the Sample failure-convention comment no longer claims 0/0 is safe; stale comments.
- **Dismissed (each documented in-code + ADR 0016 §5):** resolved-MSAA flags-lane averaging (PLAUSIBLE; two non-default session toggles; pre-existing milder variant; real fix is resolve-side, follow-up) · imported-material codec + asset-preview lanes stay Legacy (importers can't author v2; preview shader is a standalone pre-dispatch closure) · C# OLO_PROPERTY exemption (matches the AlphaMode precedent, recorded at the component).
- **Deferred follow-ups (noted, not silent):** hoist per-pixel-invariant Kulla-Conty terms out of the light loops; constexpr-decode the CPU table; check in the table generator as a tool; share test-side Hammersley/probe-harness helpers.

**Comment by drsnuggles8** [comment]
@coderabbitai review

File: OloEngine/src/OloEngine/Renderer/PathTracing/PBRClosureBSDF.h [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        glm::vec3 l;
+        if (lobeXi < pSpecular)
+        {
+            if (material.Model == PBRModel::ClosureV2)
+            {
+                // GLSL twin: closureV2SampleBRDF's specular branch — VNDF, on
+                // the same clamped roughness the pdf and evaluation use.
+                const glm::vec3 h = SampleGGXVNDF(n, v, ClosureV2Roughness(material.Roughness), xi);
+                l = glm::reflect(-v, h);
+            }
```
[the remark this answers is not in the document]
Comment (drsnuggles8) [reply OloEngine/src/OloEngine/Renderer/PathTracing/PBRClosureBSDF.h:174]: Fixed in 6be4f5b87: the ClosureV2 specular branch now returns false when dot(n, v) <= 0 before touching the VNDF sampler, with the matching Pdf<=0 termination guard added to the GLSL twin (closureV2SampleBRDF) so the two conventions stay aligned.

File: OloEngine/src/OloEngine/Renderer/ShaderBindingLayout.h [hunk 2]
```diff
@@ -168,7 +168,7 @@ namespace OloEngine
             i32 EnableLightProbes;       // Enable light probe indirect diffuse
             f32 IBLIntensity = 1.0f;     // Runtime IBL strength multiplier
             i32 AlphaMode = 0;           // 0=Opaque, 1=Mask, 2=Blend (matches AlphaMode enum)
-            i32 Pad2 = 0;
+            i32 PBRModel = 0;            // 0=Legacy, 1=ClosureV2 (matches PBRModel enum, issue #975)
```
[the remark this answers is not in the document]
Comment (drsnuggles8) [reply OloEngine/src/OloEngine/Renderer/ShaderBindingLayout.h:171]: Fixed in 6be4f5b87: added static_assert(sizeof(PBRMaterialUBO) == 144) plus the %16 alignment assert, matching CameraUBO/ModelUBO house style. Good catch — the existing CommandDispatch assert compared sizeof against GetSize(), i.e. sizeof against itself.

File: OloEngine/tests/Rendering/PathTracing/ClosureV2GpuParityTest.cpp [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+            {
+                FramebufferSpecification spec{};
+                spec.Width = kWidth;
+                spec.Height = kHeight;
+                // RGBA32F, not 16F: a half-float readback would blur the
+                // comparison to ~1e-3 relative and hide exactly the kind of
+                // small constant-factor drift this test exists to catch.
+                spec.Attachments = { FramebufferTextureFormat::RGBA32F };
+                m_OutputFB = Framebuffer::Create(spec);
+                m_Shader = Shader::Create("assets/shaders/tests/PbrClosureV2ParityProbe.glsl");
```
[the remark this answers is not in the document]
Comment (drsnuggles8) [reply OloEngine/tests/Rendering/PathTracing/ClosureV2GpuParityTest.cpp:99]: Fixed in 6be4f5b87: Draw() now ASSERT_TRUEs both m_OutputFB and m_Shader with named diagnostics before any dereference.

File: OloEngine/tests/Rendering/PathTracing/ClosureV2Test.cpp [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+                        const glm::vec3 l(std::cos(phi) * sinTheta, std::sin(phi) * sinTheta, cosTheta);
+
+                        const glm::vec3 f = ClosureV2Evaluate(n, v, l, white, 1.0f, roughness);
+                        full += static_cast<f64>((f.x + f.y + f.z) / 3.0f) * static_cast<f64>(cosTheta);
+
+                        // The exact compensation term Evaluate added for this
+                        // direction: same clamped roughness, and dot(n, v) ==
+                        // muO / dot(n, l) == cosTheta exactly for these
+                        // axis-aligned constructions.
+                        const glm::vec3 ms = ClosureV2MultiScatter(muO, cosTheta, rClamped, white);
```
[the remark this answers is not in the document]
Comment (drsnuggles8) [reply OloEngine/tests/Rendering/PathTracing/ClosureV2Test.cpp]: Fixed in 6be4f5b87: the furnace subtraction now passes the authored roughness (matching the energy-lookup contract closureV2Evaluate follows) and the unused rClamped local is gone. The tile-material scene-YAML round-trip from the outside-diff comment is in the same commit, mirroring the MaterialComponent omit-Legacy/Reject idiom on both write and read.

File: OloEngine/tests/Rendering/PathTracing/ClosureV2GpuParityTest.cpp [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            }
+
+            void Draw()
+            {
+                // Fail loudly on a resource-creation failure (a missing probe
+                // shader or an FBO the driver refused) instead of nullptr-
+                // dereferencing below, which reads as a crash rather than a
+                // diagnosable setup problem.
+                ASSERT_TRUE(m_OutputFB != nullptr) << "parity probe framebuffer was not created";
+                ASSERT_TRUE(m_Shader != nullptr) << "PbrClosureV2ParityProbe.glsl failed to load/compile";
```
[the remark this answers is not in the document]
Comment (drsnuggles8) [reply OloEngine/tests/Rendering/PathTracing/ClosureV2GpuParityTest.cpp]: Fixed in 697a47f5a: Draw() now returns a [[nodiscard]] bool (ADD_FAILURE + false on a null framebuffer/shader, true on success) and all three call sites ASSERT_TRUE it before ReadOutput() — the status propagates instead of only exiting the helper.

**Comment by drsnuggles8** [comment]
## 🤖 Self-review update @ `697a47f5a7f2195db96b11a138bef5c2dd5d4c9a`
The two commits since the original review (`6be4f5b87`, `697a47f5a`) are themselves review-driven fixes, re-reviewed and verified: round-1 findings (backfacing-view guard in the v2 sampler + GLSL twin convention, PBRMaterialUBO hard-value static_assert, harness resource asserts, authored-roughness in the furnace subtraction, tile-material PBRModel in scene YAML + save-games) and round-2's status-propagation follow-up on the harness. Targeted suites re-run green after each (210 and 3 tests respectively); CI fully green on all three rounds; 0 unresolved threads.

**Exit state:** MERGEABLE · every check SUCCESS · all review threads resolved.

**Comment by drsnuggles8** [comment]
Follow-ups from this PR's review dispositions are now filed and scored, each `blocked_by: [975]` until this merges:
- #996 — make the G-Buffer flags lane resolve-safe and multi-bit (the resolved-MSAA averaging limitation from ADR 0016 §5, plus the 1-bit model ceiling)
- #997 — hoist per-pixel-invariant Kulla-Conty terms out of the light loops + constexpr-decode the CPU energy table (bit-exact perf, deliberately under the low-value floor)
- #998 — check in the GGX energy-table generator so 'generated' has a checked-in baker
